presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -117,21 +121,39 @@ else if (operator instanceof DeleteOperator) {
         this.deleteOperator = deleteOperator;
 
         currentTaskSource = sourceOperator.map(operator -> new TaskSource(operator.getSourceId(), ImmutableSet.of(), false)).orElse(null);
+        synchronized (driverBlockedFutureLock) {
+            driverBlockedFuture = SettableFuture.create();
+            // initially the driver is not blocked
+            driverBlockedFuture.set(null);
+        }
+    }
+
+    public void initialize()
+    {
+        if (initialized.compareAndSet(false, true)) {
+            operators.stream()
+                    .map(Operator::getOperatorContext)
+                    .forEach(operatorContext -> operatorContext.setMemoryRevocationRequestListener(this::completeDriverBlockedFuture));
+        }
     }
 
     public DriverContext getDriverContext()
     {
+        checkState(initialized.get(), "Driver is not initialized");
         return driverContext;
     }
 
     public Optional<PlanNodeId> getSourceId()
     {
+        checkState(initialized.get(), "Driver is not initialized");
         return sourceOperator.map(SourceOperator::getSourceId);
     }
 
     @Override
     public void close()
     {
+        checkState(initialized.get(), "Driver is not initialized");
```

Comment:
Close is called in cleanup-after-exceptional-condition situations, so it should just do the job. If it is to complain about anything, this should be the last thing it does.